### PR TITLE
Add support for "Always on top" mode

### LINKED ISSUE #3038: Feature Request: Always On Top
# Description of the new feature/enhancement
Option to let terminal be always on top of all windows, just like the new calculator app. This would be useful for people learning how to do stuff, as it can be on top of the tutorial window while not suffering from text going on to the next line during resizing.

### PR DESCRIPTION
**Pull request by zadjii-msft**
This PR adds support for always on top mode, via two mechanisms:
* The global setting `alwaysOnTop`. When set to true, the window will be
  created in the "topmost" group of windows.  Changing this value will
  hot-reload whether the window is in the topmost group.
* The action `toggleAlwaysOnTop`, which will toggle the `alwaysOnTop`
  property at runtime.

## Detailed Description of the Pull Request / Additional comments

All "topmost" windows maintain an internal z-ordering relative to one
another, but they're all always above all other "non-topmost" windows.
So multiple Windows Terminal windows which are both `alwaysOnTop` will
maintain a z-order relative to one another, but they'll all be on top of
all other windows.

## Validation Steps Performed

Toggled always on top mode, both in the settings and also at runtime,
and verified that it largely did what I expected.

Closes #3038

### CODE REVIEW

File: src/cascadia/TerminalApp/AppLogic.idl [hunk 1]
```diff
@@ -58,7 +60,8 @@ namespace TerminalApp
         event Windows.Foundation.TypedEventHandler<Object, String> TitleChanged;
         event Windows.Foundation.TypedEventHandler<Object, LastTabClosedEventArgs> LastTabClosed;
         event Windows.Foundation.TypedEventHandler<Object, Windows.UI.Xaml.ElementTheme> RequestedThemeChanged;
-        event Windows.Foundation.TypedEventHandler<Object, ToggleFullscreenEventArgs> ToggleFullscreen;
         event Windows.Foundation.TypedEventHandler<Object, ToggleFocusModeEventArgs> ToggleFocusMode;
+        event Windows.Foundation.TypedEventHandler<Object, ToggleFullscreenEventArgs> ToggleFullscreen;
+        event Windows.Foundation.TypedEventHandler<Object, AlwaysOnTopChangedEventArgs> AlwaysOnTopChanged;
```
Comment (DHowett) [inline src/cascadia/TerminalApp/AppLogic.idl:65]: ghghgh consistency

so we're going from an event that says "flip your state" (with the two toggles) to an event that says "state changed, ask me for the new one"? should we align one or the other of the categories?

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 2]
```diff
@@ -1913,6 +1914,12 @@ namespace winrt::TerminalApp::implementation
             _UpdateTabWidthMode();
             _CreateNewTabFlyout();
         }
+
+        // Reload the current value of alwaysOnTop from the settings file. This
+        // will let the user hot-reload this setting, but any runtime changes to
+        // the alwaysOnTop setting will be lost.
+        _isAlwayOnTop = _settings->GlobalSettings().AlwaysOnTop();
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1921]: `alway`

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 3]
```diff
@@ -1976,10 +1983,8 @@ namespace winrt::TerminalApp::implementation
     // - <none>
     void TerminalPage::ToggleFocusMode()
     {
-        _toggleFocusModeHandlers(*this, nullptr);
-
         _isInFocusMode = !_isInFocusMode;
-
+        _toggleFocusModeHandlers(*this, nullptr);
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1987]: whoa why the move?

File: src/cascadia/TerminalApp/TerminalPage.idl [hunk 4]
```diff
@@ -28,8 +31,9 @@ namespace TerminalApp
         event Windows.Foundation.TypedEventHandler<Object, String> TitleChanged;
         event Windows.Foundation.TypedEventHandler<Object, LastTabClosedEventArgs> LastTabClosed;
         event Windows.Foundation.TypedEventHandler<Object, Windows.UI.Xaml.UIElement> SetTitleBarContent;
-        event Windows.Foundation.TypedEventHandler<Object, ToggleFullscreenEventArgs> ToggleFullscreen;
         event Windows.Foundation.TypedEventHandler<Object, ToggleFocusModeEventArgs> ToggleFocusMode;
+        event Windows.Foundation.TypedEventHandler<Object, ToggleFullscreenEventArgs> ToggleFullscreen;
+        event Windows.Foundation.TypedEventHandler<Object, AlwaysOnTopChangedEventArgs> AlwaysOnTopChanged;
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalPage.idl:36]: if we're never passing args we could use IInspectable for the second arg

File: src/cascadia/WindowsTerminal/IslandWindow.cpp [hunk 5]
```diff
@@ -463,6 +464,33 @@ void IslandWindow::ToggleFullscreen()
...
+    const auto hwnd = GetHandle();
+    if (hwnd)
+    {
+        const til::rectangle windowPos{ GetWindowRect() };
+        SetWindowPos(hwnd,
+                     _alwaysOnTop ? HWND_TOPMOST : HWND_NOTOPMOST,
+                     windowPos.left<int>(),
+                     windowPos.top<int>(),
+                     windowPos.width<int>(),
+                     windowPos.height<int>(),
```
Comment (DHowett) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp:489]: do we need to pass the coordinates if we're using NOSIZE?

File: src/cascadia/TerminalApp/AppLogic.idl (hunk 1 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/AppLogic.idl:65]: Eh, yes. This is now weird between the three. The "the state changed, come ask what it is" is more idiomatic, and flexible, so I'll update the others too.

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 3 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:1987]: just consistency's sake honestly

File: src/cascadia/WindowsTerminal/IslandWindow.cpp [hunk 6]
```diff
@@ -51,17 +51,18 @@ void IslandWindow::MakeWindow() noexcept
...
-                              WS_OVERLAPPEDWINDOW,
-                              CW_USEDEFAULT,
-                              CW_USEDEFAULT,
-                              CW_USEDEFAULT,
-                              CW_USEDEFAULT,
-                              nullptr,
-                              nullptr,
-                              wc.hInstance,
-                              this));
+    WINRT_VERIFY(CreateWindowEx(_alwaysOnTop ? WS_EX_TOPMOST : 0,
```
Comment (leonMSFT) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp:54]: Whoaaa, Is this all it takes to keep it topmost? That's surprisingly simple 😄

Comment (zadjii-msft) [reply src/cascadia/WindowsTerminal/IslandWindow.cpp:54]: Y U P

**Comment by carlos-zamora** [comment]
Don't forget the docs yo

**Comment by DHowett** [comment]
@msftbot merge this in 10 minutes

**Comment by DHowett** [comment]
@msftbot merge this in 1 minute

**Pull request merged**